net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild

```diff
@@ -0,0 +1,167 @@
+# Copyright 1999-2012 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+# $Header: $
+
+EAPI=5
+
+MY_PN="${PN/-drivers/}"
+
+inherit eutils autotools flag-o-matic multilib
+
+DESCRIPTION="Canon InkJet Printer Driver for Linux (Pixus/Pixma-Series)."
+HOMEPAGE="http://support-sg.canon-asia.com/contents/SG/EN/0100469302.html"
+SRC_URI="http://gdlp01.c-wss.com/gds/3/0100004693/01/${MY_PN}-source-${PV}-1.tar.gz"
+
+LICENSE="GPL-2"
+SLOT="${PV}"
+KEYWORDS="~amd64 ~x86"
+PRINTER_USE=( mp230 mg2200 e510 mg3200 mg4200 ip7200 mg5400 mg6300 )
+PRINTER_ID=( 401 402 403 404 405 406 407 408 )
+IUSE="${PRINTER_USE[@]} +servicetools"
+
+RDEPEND="
+	net-print/cnijfilter[servicetools?]
+	>=media-libs/libpng-1.5
+	>=media-libs/tiff-3.4
+	>=net-print/cups-1.4
+	servicetools? (
+		>=dev-libs/libxml2-2.7.3-r2
+		>=x11-libs/gtk+-2.6:2
+	)
+"
+DEPEND="${DEPEND}
+	sys-devel/gettext
+"
+
+REQUIRED_USE="|| ( ${PRINTER_USE[@]} )"
+
+S="${WORKDIR}/${MY_PN}-source-${PV}-1"
+
+_dir_build() {
+	local dirs=$1
+	local command=$2
+	local d
+
+	[[ $# -ne 2 ]] && die "Call as: _dir_build DIRS COMMAND"
+
+	for d in ${dirs}; do
+		local suffix=""
+		echo ">>> Working in: ${d}"
+		pushd ${d} >/dev/null
+		# progpath must be set otherwise we go for /usr/local/bin
+		if [[ ${command} == econf* ]]; then
+			${command} --enable-progpath="${EPREFIX}/usr/bin"
+		else
+			${command}
+		fi
+		popd > /dev/null
+	done
+}
+
+_printer_dir_build() {
+	local command=$1
+	local d
+
+	[[ $# -ne 1 ]] && die "Call as: _printer_dir_build COMMAND"
+
+	for (( i=0; i<${#PRINTER_USE[@]}; i++ )); do
+		local name="${PRINTER_USE[$i]}"
+		if use ${name}; then
+			for d in ${DIRS_PRINTER}; do
+				echo ">>> Working in: ${name}/${d}"
+				pushd ${name}/${d} > /dev/null
+				if [[ ${command} == econf* ]]; then
+					${command} \
+						--enable-progpath="${EPREFIX}/usr/bin" \
+						--program-suffix=${name}
+				else
+					${command}
+				fi
+				popd > /dev/null
+			done
+		fi
+	done
+}
+
+pkg_setup() {
+	[[ -z ${LINGUAS} ]] && LINGUAS="en"
+
+	DIRS_PRINTER="cnijfilter"
+	# lgmon must be first as it is required by cngpijmon
+	use servicetools && DIRS_PRINTER+=" lgmon cngpijmon maintenance"
+}
+
+src_prepare() {
+	local d i
+
+	# missing macros directory make aclocal fail
+	mkdir maintenance/m4 || die
+
+	epatch \
+		"${FILESDIR}/${MY_PN}"-3.70-png.patch \
+		"${FILESDIR}/${MY_PN}"-3.70-ppd.patch \
+		"${FILESDIR}/${MY_PN}"-3.70-ppd2.patch \
+		"${FILESDIR}/${MY_PN}"-3.70-libexec-cups.patch \
+		"${FILESDIR}/${MY_PN}"-3.70-libexec-backend.patch \
+		"${FILESDIR}/${MY_PN}"-3.80-cups.patch
+
+	_dir_build "${DIRS_PRINTER}" "eautoreconf"
+
+	for (( i=0; i<${#PRINTER_USE[@]}; i++ )); do
+		local name="${PRINTER_USE[$i]}"
+		local pid="${PRINTER_ID[$i]}"
+		if use ${name}; then
+			mkdir -p ${name} || die
+			ln -s "${S}"/${pid} ${name}/ || die
+			for d in ${DIRS_PRINTER}; do
+				cp -a ${d} ${name} || die
```

Comment:
This whole logic seems like reinventing the wheel of multibuild.eclass. If you need to build the sources for every model separately, reuse the logic already in the eclass instead of reinventing your own, poorly.

Reply:
Could you please elaborate ? I know that `multilib` is used to build variants from the same source, but I don't know how to use it to iterate over different sources (= sub-directories here).

Reply:
Look at multibuild.eclass. It's used to build N different copies (variants) of the same package.

Reply:
OK, but in that case, we don't want to build variants of the same package. We have several subdirectories in the source package, each of them is a component having its own configure and makefile, so we have to iterate over that subdirectories (and select some of them depending on USE flags). You wrote that the helper function _dir_build() is ugly, but what do you suggest as a cleaner way to iterate over that subdirectories ?

Reply:
But in this step, you are creating a copy of every input directory for every printer enabled, am I correct? This is what multibuild is for.

For using those directories afterwards, just inline a trivial for loop, without any fancy prints.

Reply:
By the way, try to build out-of-source if possible; i.e. instead of copying those directories use build dirs with `ECONF_SOURCE=${S}/${dir} econf`.

Reply:
OK, got it ! Yes, this source copy inherited from Sabayon is ugly and I will try to improve that part when reworking the `cnijfilter-drivers` package for the second commit. My question is more on the current `cnijfilter` package and its _dir_build() helper. What do you suggest to make it cleaner than in my latest intermediate commit ?

Reply:
Closed as TODO in `cnijfilter-drivers` ebuild.